src/ui/carousel/carousel-synchronizer.ts

```diff
@@ -0,0 +1,173 @@
+import {Carousel, CarouselEvent} from './carousel';
+import { DefaultMap } from '../../map/default-map';
+
+/**
+ * Will synchronize the active slide index across multiple carousels.
+ *
+ * This can be useful if you have separate areas for assets, copy and/or
+ * navigation that need transition together.
+ *
+ * Example:
+ *
+ * DOM:
+ * ```
+ * <div class="carousel">
+ *   <div class="slide">1</div>
+ *   <div class="slide">2</div>
+ *   <div class="slide">3</div>
+ * </div>
+ * <div class="nav">
+ *   <div class="dot">1</div>
+ *   <div class="dot">2</div>
+ *   <div class="dot">3</div>
+ * </div>
+ * ```
+ *
+ * JS:
+ * ```
+ * const a = new Carousel(
+ *   document.querySelector('.carousel'),
+ *   Array.from(document.querySelectorAll('.container-a .slide')),
+ *   {
+ *     allowLooping: true,
+ *     transition: new DraggableSlide(),
+ *   });
+ * const b = new Carousel(
+ *   document.querySelector('.container-b'),
+ *   Array.from(document.querySelectorAll('.container-b .slide')),
+ *   {
+ *     allowLooping: true,
+ *   });
+ * new CarouselSynchronizer(a, b);
+ *
+ * // At some point later, dispose.
+ * a.dispose();
+ * b.dispose();
+ * syncInstance.dispose(a);
+ * ```
+ *
+ * CSS:
+ * ```
+ * .carousel {
+ *   position: relative;
+ *   height: 20vh;
+ *   width: 80vw;
+ *   overflow: hidden;
+ *   display: flex;
+ *   flex-wrap: nowrap;
+ * }
+ * .slide {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ *   height: 100%;
+ *   width: 50%;
+ *   flex-shrink: 0;
+ *   font-size: 10vw;
+ *   user-select: none;
+ * }
+ * .slide:nth-child(1) {
+ *   background: paleturquoise;
+ * }
+ * .slide:nth-child(2) {
+ *   background: palevioletred;
+ * }
+ * .slide:nth-child(3) {
+ *   background: palegreen;
+ * }
+ * .nav {
+ *   display: flex;
+ *   justify-content: center;
+ *   align-items: center;
+ * }
+ * .dot {
+ *   flex-grow: 0;
+ *   flex-shrink: 0;
+ *   width: 10px;
+ *   height: 10px;
+ *   border-radius: 50%;
+ *   opacity: .25;
+ *   background: black;
+ *   transition: opacity 1s, transform 1s;
+ *   margin: 0 8px;
+ * }
+ * .dot.active {
+ *   transform: scale(1.25);
+ *   opacity: 1;
+ * }
+ * .dot.before, .dot.after {
+ *   transform: scale(0.9);
+ * }
+ * ```
+ */
+export class CarouselSynchronizer {
+  private graph: DefaultMap<Carousel, Set<Carousel>>;
+
+  constructor(...carousels: Carousel[]) {
+    this.graph = DefaultMap.usingFunction((carousel: Carousel) => new Set());
+    this.sync(...carousels);
```

Comment:
Maybe it's just the way I think about how object creation should work, but in my mind the constructor should basically set up the attributes for the class and shouldn't necessarily do the actual work. So I'd suggest the example usage of this class be something like:

```
const syncer = new CarouselSynchronizer(); 
syncer.sync(a, b); 
```

I think this helps readability because it tells me "line 1: create a syncer object, line 2: sync carousels a and b". (Whereas when I read `new CarouselSynchronizer(a, b)` I get confused about what's going on.)

(tl;dr: I'm suggesting removing the `carousels` args in the constructor.)

Reply:
I'm good with that.